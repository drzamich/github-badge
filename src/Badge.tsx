import React from 'react';
import styled from 'styled-components';
import { User } from './interfaces';
import { Color, Heading } from './styles';

interface ContainerDiv {
  $loading: boolean,
}

const Container = styled.div<ContainerDiv>`
  display: ${(props) => (props.$loading ? 'none' : 'block')};
`;

const IdContainer = styled.div`
  width: 100%;
  display: flex;
  align-items: flex-end;
`;

const Img = styled.img`
  width: 64px;
  height: 64px;
  border-radius: 12px;
`;

const Username = styled.h3`
  word-spacing: 300px;
  margin-left: 16px;
`;

const Bio = styled.p`
  color: ${Color.Gray};
  margin-top: 16px;
`;

const TopReposHeading = styled(Heading)`
  margin: 16px 0;
`;

const Repos = styled.ul`
  list-style: none;
  a {
    display: block;
    padding: 16px;
    box-shadow: 0px 2px 7px rgba(0, 0, 0, 0.1);
    border-radius: 12px;
    color: ${Color.RoyalBlue};
    text-decoration: none;
    margin-bottom: 16px;
    word-break: break-all;

    &:hover {
      text-decoration: underline;
    }
  }
`;

interface BadgeProps {
  user: User | undefined,
  loading: boolean,
  setLoading: React.Dispatch<React.SetStateAction<boolean>>,
}

export const Badge: React.FC<BadgeProps> = ({ user, loading, setLoading }) => (
  <Container $loading={loading} aria-hidden={loading}>
    <IdContainer>
      <Img
        src={user?.avatar_url}
        alt={user?.name || user?.login}
        onLoad={() => setLoading(false)}
      />
      <Username>{user?.name || user?.login}</Username>
    </IdContainer>

    {user?.bio && <Bio>{user?.bio}</Bio>}
    {!(user?.repos?.length) && <TopReposHeading>User has no repositories.</TopReposHeading> }
    {!!(user?.repos?.length)
  && (
    <>
      <TopReposHeading>Top repositories</TopReposHeading>
      <Repos>
        {user?.repos.map((repo) => (
          <li key={repo.html_url}>
            <a href={repo.html_url}>{repo.full_name}</a>
          </li>
        ))}
      </Repos>
    </>
  ) }
  </Container>
);
