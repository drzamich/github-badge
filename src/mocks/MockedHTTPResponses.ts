import { AxiosError, AxiosResponse } from 'axios';
import { user as ValidUserResponse, repos as ValidReposResponse } from './ValidAPIResponses.json';
import { Repository, User } from '../interfaces';

const defResponse = {
  headers: [],
  config: {},
};

const defSuccessResponse = {
  ...defResponse,
  status: 200,
  statusText: 'success',
};

const defErrorResponse: AxiosResponse = {
  ...defResponse,
  status: 500,
  statusText: 'success',
  data: null,
};

const defError: AxiosError = {
  config: {},
  isAxiosError: true,
  toJSON: () => ({}),
  name: '',
  message: '',
};

export const userReponse: AxiosResponse<User> = {
  data: ValidUserResponse,
  ...defSuccessResponse,
};

export const reposResponse: AxiosResponse<Repository[]> = {
  data: ValidReposResponse,
  ...defSuccessResponse,
};

export const networkError: AxiosError = {
  ...defError,
  request: {},
};

export const internalServerError: AxiosError = {
  ...defError,
  response: {
    ...defErrorResponse,
  },
};

export const notFoundError: AxiosError = {
  ...defError,
  response: {
    ...defErrorResponse,
    status: 404,
  },
};
