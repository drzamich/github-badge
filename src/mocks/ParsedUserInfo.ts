/* eslint-disable max-len */
import { User, Repository } from '../interfaces';

export const parsedUser: User = {
  name: 'Tejas Kumar',
  login: 'TejasQ',
  avatar_url: 'https://avatars1.githubusercontent.com/u/9947422?v=4',
  bio: 'I write code and love life.',
};

export const parsedRepos: Repository[] = [
  { full_name: 'TejasQ/add-gitignore', html_url: 'https://github.com/TejasQ/add-gitignore', stargazers_count: 450 },
  { full_name: 'TejasQ/anna-artemov.now.sh', html_url: 'https://github.com/TejasQ/anna-artemov.now.sh', stargazers_count: 5 },
  { full_name: 'TejasQ/awesome-cli-apps', html_url: 'https://github.com/TejasQ/awesome-cli-apps', stargazers_count: 3 },
];

export const finalUser: User = {
  ...parsedUser,
  repos: parsedRepos,
};
