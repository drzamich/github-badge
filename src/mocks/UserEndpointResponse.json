{
  "login": "TejasQ",
  "id": 9947422,
  "node_id": "MDQ6VXNlcjk5NDc0MjI=",
  "avatar_url": "https://avatars1.githubusercontent.com/u/9947422?v=4",
  "gravatar_id": "",
  "url": "https://api.github.com/users/TejasQ",
  "html_url": "https://github.com/TejasQ",
  "followers_url": "https://api.github.com/users/TejasQ/followers",
  "following_url": "https://api.github.com/users/TejasQ/following{/other_user}",
  "gists_url": "https://api.github.com/users/TejasQ/gists{/gist_id}",
  "starred_url": "https://api.github.com/users/TejasQ/starred{/owner}{/repo}",
  "subscriptions_url": "https://api.github.com/users/TejasQ/subscriptions",
  "organizations_url": "https://api.github.com/users/TejasQ/orgs",
  "repos_url": "https://api.github.com/users/TejasQ/repos",
  "events_url": "https://api.github.com/users/TejasQ/events{/privacy}",
  "received_events_url": "https://api.github.com/users/TejasQ/received_events",
  "type": "User",
  "site_admin": false,
  "name": "Tejas Kumar",
  "company": "@g2i",
  "blog": "https://twitter.com/tejaskumar_",
  "location": "Berlin, DE",
  "email": null,
  "hireable": true,
  "bio": "I write code and love life.",
  "twitter_username": null,
  "public_repos": 60,
  "public_gists": 1,
  "followers": 753,
  "following": 280,
  "created_at": "2014-11-25T13:12:53Z",
  "updated_at": "2020-11-25T21:05:04Z"
}