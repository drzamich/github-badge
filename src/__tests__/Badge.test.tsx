import React from 'react';
import { render, screen } from '@testing-library/react';
import { Badge } from '../Badge';
import { finalUser } from '../mocks/ParsedUserInfo';

describe('Badge', () => {
  const setLoading = jest.fn();

  it('Renders user info properly when loading is completed', () => {
    // Given
    const user = finalUser;
    const loading = false;
    // When
    const badge = render(<Badge user={user} loading={loading} setLoading={setLoading} />);
    // Then
    expect(screen.getByText('Top repositories')).toBeVisible();
    expect(badge).toMatchSnapshot();
  });

  it('Does not render user info when loading is not completed', () => {
    // Given
    const user = finalUser;
    const loading = true;
    // When
    const badge = render(<Badge user={user} loading={loading} setLoading={setLoading} />);
    // Then
    expect(screen.getByText('Top repositories')).not.toBeVisible();
    expect(badge).toMatchSnapshot();
  });
});
