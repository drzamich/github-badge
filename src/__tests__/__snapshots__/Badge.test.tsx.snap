// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Does not render user info when loading is not completed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-hidden="true"
        class="sc-dlfnbm iTRrSd"
      >
        <h3
          class="sc-gsTCUz hmKoqn"
        >
          GitHub user found.
        </h3>
        <div
          class="sc-hKgILt oomuw"
        >
          <img
            alt="Avatar of Tejas Kumar"
            class="sc-eCssSg bhbAo"
            src="https://avatars1.githubusercontent.com/u/9947422?v=4"
          />
          <h3
            class="sc-jSgupP bLJygP"
          >
            Tejas Kumar
          </h3>
        </div>
        <p
          class="sc-gKsewC hnFWpo"
        >
          I write code and love life.
        </p>
        <h3
          class="sc-bdfBwQ sc-iBPRYJ fVVSDW guCgMS"
        >
          Top repositories
        </h3>
        <ul
          class="sc-fubCfw eZIGwe"
        >
          <li>
            <a
              href="https://github.com/TejasQ/add-gitignore"
            >
              TejasQ/add-gitignore
            </a>
          </li>
          <li>
            <a
              href="https://github.com/TejasQ/anna-artemov.now.sh"
            >
              TejasQ/anna-artemov.now.sh
            </a>
          </li>
          <li>
            <a
              href="https://github.com/TejasQ/awesome-cli-apps"
            >
              TejasQ/awesome-cli-apps
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-hidden="true"
      class="sc-dlfnbm iTRrSd"
    >
      <h3
        class="sc-gsTCUz hmKoqn"
      >
        GitHub user found.
      </h3>
      <div
        class="sc-hKgILt oomuw"
      >
        <img
          alt="Avatar of Tejas Kumar"
          class="sc-eCssSg bhbAo"
          src="https://avatars1.githubusercontent.com/u/9947422?v=4"
        />
        <h3
          class="sc-jSgupP bLJygP"
        >
          Tejas Kumar
        </h3>
      </div>
      <p
        class="sc-gKsewC hnFWpo"
      >
        I write code and love life.
      </p>
      <h3
        class="sc-bdfBwQ sc-iBPRYJ fVVSDW guCgMS"
      >
        Top repositories
      </h3>
      <ul
        class="sc-fubCfw eZIGwe"
      >
        <li>
          <a
            href="https://github.com/TejasQ/add-gitignore"
          >
            TejasQ/add-gitignore
          </a>
        </li>
        <li>
          <a
            href="https://github.com/TejasQ/anna-artemov.now.sh"
          >
            TejasQ/anna-artemov.now.sh
          </a>
        </li>
        <li>
          <a
            href="https://github.com/TejasQ/awesome-cli-apps"
          >
            TejasQ/awesome-cli-apps
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Badge Renders user info properly when loading is completed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        aria-hidden="false"
        class="sc-dlfnbm gIdFjC"
      >
        <h3
          class="sc-gsTCUz hmKoqn"
        >
          GitHub user found.
        </h3>
        <div
          class="sc-hKgILt oomuw"
        >
          <img
            alt="Avatar of Tejas Kumar"
            class="sc-eCssSg bhbAo"
            src="https://avatars1.githubusercontent.com/u/9947422?v=4"
          />
          <h3
            class="sc-jSgupP bLJygP"
          >
            Tejas Kumar
          </h3>
        </div>
        <p
          class="sc-gKsewC hnFWpo"
        >
          I write code and love life.
        </p>
        <h3
          class="sc-bdfBwQ sc-iBPRYJ fVVSDW guCgMS"
        >
          Top repositories
        </h3>
        <ul
          class="sc-fubCfw eZIGwe"
        >
          <li>
            <a
              href="https://github.com/TejasQ/add-gitignore"
            >
              TejasQ/add-gitignore
            </a>
          </li>
          <li>
            <a
              href="https://github.com/TejasQ/anna-artemov.now.sh"
            >
              TejasQ/anna-artemov.now.sh
            </a>
          </li>
          <li>
            <a
              href="https://github.com/TejasQ/awesome-cli-apps"
            >
              TejasQ/awesome-cli-apps
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-hidden="false"
      class="sc-dlfnbm gIdFjC"
    >
      <h3
        class="sc-gsTCUz hmKoqn"
      >
        GitHub user found.
      </h3>
      <div
        class="sc-hKgILt oomuw"
      >
        <img
          alt="Avatar of Tejas Kumar"
          class="sc-eCssSg bhbAo"
          src="https://avatars1.githubusercontent.com/u/9947422?v=4"
        />
        <h3
          class="sc-jSgupP bLJygP"
        >
          Tejas Kumar
        </h3>
      </div>
      <p
        class="sc-gKsewC hnFWpo"
      >
        I write code and love life.
      </p>
      <h3
        class="sc-bdfBwQ sc-iBPRYJ fVVSDW guCgMS"
      >
        Top repositories
      </h3>
      <ul
        class="sc-fubCfw eZIGwe"
      >
        <li>
          <a
            href="https://github.com/TejasQ/add-gitignore"
          >
            TejasQ/add-gitignore
          </a>
        </li>
        <li>
          <a
            href="https://github.com/TejasQ/anna-artemov.now.sh"
          >
            TejasQ/anna-artemov.now.sh
          </a>
        </li>
        <li>
          <a
            href="https://github.com/TejasQ/awesome-cli-apps"
          >
            TejasQ/awesome-cli-apps
          </a>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
